doc: chip_tool_guide: doorlock user & credentials (#27601)
project-chip/connectedhomeip: docs/guides/chip_tool_guide.md

    ```
    ./chip-tool any read-by-id 0xFFFFFFFF 0xFFFFFFFF 1 0xFFFF
    ```

<hr>

> ## Saving users and credentials on door lock devices

Reviewer: Should this be in the main chip-tool guide?  Or should it be a separate guide that is linked from here?  Seems like we might want some per-complicated-device guides...

Author: Fair point. I suggest we wait until there are more sections like this and then start creating device type guides, because it will affect more than just the CHIP Tool Guide. Separating this content from the CHIP Tool Guide now could lead to having two CHIP Tool guides in parallel. 

Would you have suggestions what types of devices might require separate guides?

Reviewer: Thermostats seem like an obvious one that might be complicated.

And maybe bridges.

Matter door lock devices can store pools of users and credentials that allow you to configure different access scenarios. Each user and credential in the pool has an index value. Additionally, each user has a list of Occupied Credentials.

By default, each door lock device comes with no user or credential defined, but it reserves several slots in both pools that can be populated with new users and credentials, up to the value specified in the the `NumberOfTotalUsersSupported` attribute and the `NumberOfCredentialsSupportedPerUser` attribute, respectively.

> For security reasons, all communication between users and credentials happens only using their respective index values. No other information is shared between both pools.

Reviewer: I don't think it's "for security reasons".  It just is.

The CHIP Tool lets you add users and credentials on door lock devices and securely match their indexes to one another. This is an optional feature available only when working with devices that implement the `doorlock` cluster.

> **Note:**
> Users and credentials can only be modified by whoever has the right permissions, as specified in the Access Control List.

> Complete the following steps:

Reviewer 2: This sentence looks like taken out of context. Maybe it is worth adding something like: `To save credentials and users complete the following steps.`

1. Set up a user on the device.
1. Assign a credential for the newly created user.

### Step 1: Set up a user

To set up a user on a door lock device with the CHIP Tool, use the following command pattern:

```

> $ ./chip-tool doorlock set-user <OperationType> <UserIndex> <UserName> <UserUniqueID> <UserStatus> <UserType> <CredentialRule> <destination-id> <endpoint-id> --timedInteractionTimeoutMs 1000

Reviewer: Why the hardcoded 1000?  The actual number here depends on what your goals are.

```

In this command:

-   _<OperationType\>_ is one of the available types of operation for the user:

    - `Add` - This operation adds a new user to the slot at _<UserIndex\>_.
    - `Clear` - This operation removes an existing user from the slot at _<UserIndex\>_.
    - `Modify` - This operation modifies an existing user at the slot at _<UserIndex\>_.

-   _<UserIndex\>_ is the index value of the user, between `1` and the maximum value you set in the `NumberOfTotalUsersSupported` attribute. Setting the user index to `0` will cause an error.
-   _<UserName\>_ is the name of the user, which can have maximum 10 bytes of size.
-   _<UserUniqueID\>_ is a 4-byte number that describes the unique user ID.
-   _<UserStatus\>_ can be set to one of the following values:

>     - Available (`0`) - This status indicates that the given user slot is available for modification on the device.

Reviewer: Note that you cannot "Add" but set to "Available"....  And there are various other constraints here too.

Also, various of these things are nullable; UserUniqueID can be null.  The above text makes it sound like an id is required.

>     - Occupied (`1`) - This status indicates that the given user slot is used and active.

Reviewer 2: I've compared it with the spec and I can see that there are 3 options - 
0 as Available, which is mandatory,
1 as OccupiedEnabled, which is mandatory
3 as OccupiedDisabled, which is optional. 
I'm wondering whether it is worth adding the third option here or not because in real case the user wants to set 0  as available or 1 as occupied.

-   _<CredentialRule\>_ is the number of credentials that must be used to unlock the door lock. This parameter can have the following values:

    - `0` (Single) - One credential type is required to unlock.
    - `1` (Dual) - Two credential types are required to unlock.
    - `2` (Tri) - Three credential types are required to unlock.

-   _<destination-id\>_ is the ID of the door lock device.
-   _<endpoint-id\>_ is the ID of the endpoint on the door lock device.
-   `--timedInteractionTimeoutMs` is the timeout duration for the interaction, in milliseconds. This should allow enough time for the interaction to complete.

**Example of command:**

```

> $ ./chip-tool doorlock set-user 0 1 AAA 6452 1 0 0 1 1 --timedInteractionTimeoutMs 1000

Reviewer: The simplest way to add a user is to SetUser with Add, the index, and everything else null....

```

### Step 2: Assign a credential

Once you have a user created on the door lock device, use the following command pattern to assign a credential to it:

```
$ ./chip-tool doorlock set-credential <OperationType> <{Credential}> <CredentialData> <UserIndex> <UserStatus> <UserType> <destination-id> <endpoint-id> --timedInteractionTimeoutMs 1000
```
In this command:

-   _<OperationType\>_ is one of the available types of operation for the credential:

    - `Add` - This operation adds a new credential to a user at the slot at _<UserIndex\>_.
    - `Clear` - This operation removes an existing credential from the user at the slot at _<UserIndex\>_.
    - `Modify` - This operation modifies an existing credential for the user at the slot at _<UserIndex\>_.

> -   _<{Credential}\>_ is a JSON parameter with two fields:

Reviewer 2: Shall we add here info that it should be a JSON format parameter instead of providing a path to a JSON file to avoid future questions from users? :)

> **Note:** Users and credentials can only be modified by whoever has the right
> permissions, as specified in the Access Control List.

To save credentials and users, you need to complete the following steps,
described in detail in the following sections:

1. Set up a user on the device.
1. Assign a credential for the newly created user.

### Step 1: Set up a user

To set up a user on a door lock device with the CHIP Tool, use the following
command pattern:

```
$ ./chip-tool doorlock set-user <operation-type> <user-index> <user-name> <user-unique-id> <user-status> <user-type> <credential-rule> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>
```

In this command:

-   _<operation-type\>_ is one of the available types of operation for the user:

    -   `Add` - This operation sets a new user in the slot at _<user-index\>_.
    -   `Clear` - This operation removes an existing user from the slot at
        _<user-index\>_.
    -   `Modify` - This operation modifies an existing user at the slot at
        _<user-index\>_.

-   _<user-index\>_ is the index value of the user, between `1` and the maximum

>     value you set in the `NumberOfTotalUsersSupported` attribute. Setting the

Reviewer: You don't set NumberOfTotalUsersSupported... it's readonly.

    user index to `0` will cause an error.
-   _<user-name\>_ is the name of the user, which can have maximum 10 bytes of
    size. Can be set to `null`.
-   _<user-unique-id\>_ is a 4-byte number that describes the unique user ID.
    Can be set to `null`.
-   _<user-status\>_ can be set to `null` or to one of the following values:

    -   `0` (`Available`) - This status indicates that the given user slot is

>         available for modification on the device. This value is selected is you
>         set `null` for the variable.

Reviewer: "if you" instead of "is you"?

More importantly, this statement is just wrong. You can set UserStatus to null to do whatever the spec says for null. For an add, null means OccupiedEnabled. For a Modify null means "do not change the status".
```suggestion
        available for modification on the device.
```

    -   `1` (`OccupiedEnabled`) - This status indicates that the given user slot
        is used and active.
    -   `3` (`OccupiedDisabled`) - This status indicates that the given user
        slot is used, but disabled. Unlike `0` and `1`, supporting this status
        is optional.

-   _<credential-rule\>_ is the number of credentials that must be used to
    unlock the door lock. This parameter can be set to `null` or to one of the
    following values:

    -   `0` (Single) - One credential type is required to unlock. This value is

>         selected is you set `null` for the variable.

Reviewer: Not for a Modify it's not.

    -   `1` (Dual) - Two credential types are required to unlock.
    -   `2` (Triple) - Three credential types are required to unlock.

> -   _<destination-id\>_ is the ID of the door lock device.

Reviewer: ```suggestion
-   _<destination-id\>_ is the node ID of the door lock device.
```

-   _<endpoint-id\>_ is the ID of the endpoint on the door lock device.
-   `--timedInteractionTimeoutMs` is the timeout duration for the interaction,
    in milliseconds (_<ms_value>_). This should allow enough time for the

>     interaction to complete.

Reviewer: No, this is the timeout within which the _client promises to send the request_.  The interaction absolutely does not have to complete during this timeout.

Author: We opted for the following edit:
```suggestion
-   `--timedInteractionTimeoutMs` is the duration in milliseconds (_<ms_value>_)
    of the time window for receiving a request on the server side.
    This should allow enough time for receiving the request.
```

Reviewer: That seems fine.

**Examples of command:**

```

> $ ./chip-tool doorlock set-user 0 1 AAA 6452 1 0 0 1 1 --timedInteractionTimeoutMs 1000

Reviewer: This should probably describe what this command will do.  As should the next example.

```

```
$ ./chip-tool doorlock set-user 0 1 null null null null null 1 1 --timedInteractionTimeoutMs 1000
```

### Step 2: Assign a credential

Once you have a user created on the door lock device, use the following command
pattern to assign a credential to it:

```
$ ./chip-tool doorlock set-credential <operation-type> <{Credential}> <credential-data> <user-index> <user-status> <user-type> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>
```

In this command:

-   _<operation-type\>_ is one of the available types of operation for the
    credential:

    -   `Add` - This operation adds a new credential to a user at the slot at
        _<user-index\>_.
    -   `Clear` - This operation removes an existing credential from the user at
        the slot at _<user-index\>_.
    -   `Modify` - This operation modifies an existing credential for the user
        at the slot at _<user-index\>_.

> -   _<{Credential}\>_ is a parameter in the JSON format, with the following two

Reviewer: ```suggestion
-   _<{Credential}\>_ is a parameter in the JSON format, with the following two
```
It's a JSON _object_ to be precise, right?  Sticking a JSON array or string or number in here will not work.

    fields:

>     -   _<credential-type\>_ is the type of the credential. It can have one of

Reviewer: What is the name of the field?  This documentation does not say....

        the following values:

        -   `0` - Programming PIN
        -   `1` - PIN
        -   `2` - RFID
        -   `3` - Fingerprint
        -   `4` - Finger vein

>     -   _<credential-index\>_ is the index of the credential, between `1` and

Reviewer: Again, what is the name of the field?

>         the maximum value you set in the `NumberOfCredentialsSupportedPerUser`

Reviewer: "You" the person operating chip-tool do not set NumberOfCredentialsSupportedPerUser. The device does.

        attribute (see the section 5.2.3.20 of the Matter Application Clusters
        specification for details). Setting the credential index to `0` will

>         cause an error.

Reviewer: Actually, if your type is ProgrammingPIN then _not_ setting credential index to 0 will cause an error...

> -   _<credential-data\>_ is a string parameter with the secret credential data.

Reviewer: It's an octet string, not a string.  You can specify it as a string for really simple cases, but you can also specify it as hex, etc.

    For example, the PIN code value (`12345` in the example below).

> -   _<user-index\>_ is the index of the user that will be associated with the
>     credential. Can be set to `null`.

Reviewer: ```suggestion
-   _<user-index\>_ is the index of the user that will be associated with the
    credential. Can be set to `null` to create a new user.
```

-   _<user-status\>_ is the status of the user that will be associated with the
    credential. See the description of this parameter in
    [Set up a user](#step-1-set-up-a-user). Can be set to `null`.

> -   _<destination-id\>_ is the ID of the door lock device.

Reviewer: ```suggestion
-   _<destination-id\>_ is the node ID of the door lock device.
```

-   _<endpoint-id\>_ is the ID of the endpoint on the door lock device.

> -   `--timedInteractionTimeoutMs` is the timeout duration for the interaction,

Reviewer: See comments above.

    in milliseconds (_<ms_value>_). This should allow enough time for the
    interaction to complete.

**Example of command:**

```
$ ./chip-tool doorlock set-credential 0 '{ "credentialType": 1, "credentialIndex": 1 }' "12345" 1 null null 1 1 --timedInteractionTimeoutMs 1000
```

### Operations on users and credentials

After you set up users and credentials on your door lock device, you can use
several CHIP Tool commands to interact with them.

All commands reuse the parameters explained earlier in this section. The
following command patterns are available:

-   Reading the status of the user:

    ```

>     $ ./chip-tool doorlock get-user <user-index> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>

Reviewer: Note that get-user does not need `--timedInteractionTimeoutMs` at all.  Take it out here, please.
```suggestion
    $ ./chip-tool doorlock get-user <user-index> <destination-id> <endpoint-id>
```

    ```

    This command returns the status of the user at the specified _<user-index\>_
    at the specified _<destination-id\>_ and _<endpoint-id\>_.

-   Reading the status of the credential:

    ```

>     $ ./chip-tool doorlock get-credential-status <{Credential}> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>

Reviewer: ```suggestion
    $ ./chip-tool doorlock get-credential-status <{Credential}> <destination-id> <endpoint-id>
```

    ```

    This command returns the status of the credential of the specified
    _<{Credential}\>_ at the specified _<destination-id\>_ and _<endpoint-id\>_.

-   Cleaning the user:

    ```
    $ ./chip-tool doorlock clear-user <user-index> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>
    ```

    This command cleans the slot containing the specified _<user-index\>_ at the
    specified _<destination-id\>_ and _<endpoint-id\>_.

-   Cleaning the credential:

    ```
    $ ./chip-tool doorlock clear-credential <{Credential}> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>
    ```

    This command cleans the slot containing the specified _<{Credential}\>_ at
    the specified _<destination-id\>_ and _<endpoint-id\>_.

### Operations with user PIN code

If you set the _<credential-type\>_ to PIN when
[assigning credentials](#step-2-assign-a-credential), you can use the following
command patterns to verify if it works and invoke it to open or close the door
lock:

-   Verifying the PIN code:

    ```

>     $ ./chip-tool doorlock read require-pinfor-remote-operation <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>

Reviewer: ```suggestion
    $ ./chip-tool doorlock read require-pinfor-remote-operation <destination-id> <endpoint-id>
```

you seem to be pretty confused about this `--timedInteractionTimeoutMs` parameter...

command pattern:

```
$ ./chip-tool doorlock set-user <operation-type> <user-index> <user-name> <user-unique-id> <user-status> <user-type> <credential-rule> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>
```

In this command:

-   _<operation-type\>_ is one of the available types of operation for the user:

    -   `Add` - This operation sets a new user in the slot at _<user-index\>_.
    -   `Clear` - This operation removes an existing user from the slot at
        _<user-index\>_.
    -   `Modify` - This operation modifies an existing user at the slot at
        _<user-index\>_.

-   _<user-index\>_ is the index value of the user, between `1` and the
    value of the `NumberOfTotalUsersSupported` attribute. Setting the
    user index to `0` will cause an error.
-   _<user-name\>_ is the name of the user, which can have maximum 10 bytes of
    size. Can be set to `null`.
-   _<user-unique-id\>_ is a 4-byte number that describes the unique user ID.
    Can be set to `null`.
-   _<user-status\>_ can be set to `null` or to one of the following values:

    -   `0` (`Available`) - This status indicates that the given user slot is

>         available for modification on the device.

Reviewer: This is not actually a valid value for the UserStatus in SetUser....

    -   `1` (Dual) - Two credential types are required to unlock.
    -   `2` (Triple) - Three credential types are required to unlock.

-   _<destination-id\>_ is the node ID of the door lock device.
-   _<endpoint-id\>_ is the ID of the endpoint on the door lock device.
-   `--timedInteractionTimeoutMs` is the duration in milliseconds (_<ms_value>_)
    of the time window for receiving a request on the server side.
    This should allow enough time for receiving the request.

**Examples of command:**

The following command runs the `set-user` command that adds (`0`) a user at the index `1`;
the user has the name `AAA` and the unique ID `6452`. The user's status is set to `OccupiedEnabled`
(`1`), the user type is set to `UnrestrictedUser` (`0`), the credential rule is set to single (`0`),
the targeted node ID of the destination door lock device is `1` and the targeted `doorlock`
cluster's endpoint ID on that device is `1`. The `--timedInteractionTimeoutMs` has a custom value.

```
$ ./chip-tool doorlock set-user 0 1 AAA 6452 1 0 0 1 1 --timedInteractionTimeoutMs 1000
```

The following command mirrors the action of the command above, but it targets an empty user name

> (`null`) with the `0xFFFF` user unique ID (`null`). The user status defaults to `OccupiedEnabled`,

Reviewer: ```suggestion
(`null`) and has `null` for the unique ID. The user status defaults to `OccupiedEnabled`,
```
There is no 0XFFFF here anywhere.

### Step 2: Assign a credential

Once you have a user created on the door lock device, use the following command
pattern to assign a credential to it:

```
$ ./chip-tool doorlock set-credential <operation-type> <{Credential}> <credential-data> <user-index> <user-status> <user-type> <destination-id> <endpoint-id> --timedInteractionTimeoutMs <ms_value>
```

In this command:

-   _<operation-type\>_ is one of the available types of operation for the
    credential:

    -   `Add` - This operation adds a new credential to a user at the slot at
        _<user-index\>_.
    -   `Clear` - This operation removes an existing credential from the user at
        the slot at _<user-index\>_.
    -   `Modify` - This operation modifies an existing credential for the user
        at the slot at _<user-index\>_.

-   _<{Credential}\>_ is a JSON object, with the following two
    fields:

    -   `"credentialType"` is the key field for the type of the credential.
        It can have one of the following values:

        -   `0` - Programming PIN
        -   `1` - PIN
        -   `2` - RFID
        -   `3` - Fingerprint
        -   `4` - Finger vein

>     -   `"credentialIndex"` is the key field for the index of the credential, between `1` and
>         the value of the `NumberOfCredentialsSupportedPerUser`
>         attribute (see the section 5.2.3.20 of the Matter Application Clusters
>         specification for details). `0` is required for the Programming PIN.
>         In other cases, setting the credential index to `0` will
>         cause an error.

Reviewer: ```suggestion
    -   `"credentialIndex"` is the key field for the index of the credential. If credentialType
        is not "Programming PIN", credentialIndex must be between `1` and
        the value of the `NumberOfCredentialsSupportedPerUser`
        attribute (see the section 5.2.3.20 of the Matter Application Clusters
        specification for details). `0` is required for the Programming PIN.
        In other cases, setting the credential index to `0` will
        cause an error.
```
